rsl/rsl.go

```diff
@@ -39,37 +43,132 @@ func InitializeNamespace() error {
+// of the entry it is revoking.
+func (r *RSLEntry) Commit(sign bool) error {
 	repo, err := common.GetRepositoryHandler()
 	if err != nil {
 		return err
 	}
-	message := fmt.Sprintf("%s: %s", refName, commitID.String())
 
-	return gitinterface.Commit(repo, plumbing.ZeroHash, RSLRef, message, sign)
+	for _, revoked := range r.Revoke {
+		revokedEntry, err := GetEntry(revoked)
+		if err != nil {
+			return err
+		}
+
+		if revokedEntry.RefName != r.RefName {
+			return ErrRSLRevokeEntryHasInvalidRef
+		}
+	}
+
+	return gitinterface.Commit(repo, plumbing.ZeroHash, RSLRef, r.createCommitMessage(), sign)
+}
+
+func (r *RSLEntry) createCommitMessage() string {
+	lines := []string{
+		fmt.Sprintf("ref: %s", r.RefName),
+		fmt.Sprintf("commit: %s", r.CommitID.String()),
+	}
+
+	for _, revoked := range r.Revoke {
+		lines = append(lines, fmt.Sprintf("revoke: %s", revoked.String()))
+	}
+	return strings.Join(lines, "\n")
 }
 
 // GetLatestEntry returns the latest entry available locally in the RSL.
 // TODO: There is no information yet about the signature for the entry.
-func GetLatestEntry() (string, plumbing.Hash, error) {
+func GetLatestEntry() (*RSLEntry, error) {
 	repo, err := common.GetRepositoryHandler()
 	if err != nil {
-		return "", plumbing.ZeroHash, err
+		return &RSLEntry{}, err
 	}
 
 	ref, err := repo.Reference(plumbing.ReferenceName(RSLRef), true)
 	if err != nil {
-		return "", plumbing.ZeroHash, err
+		return &RSLEntry{}, err
 	}
 
 	commitObj, err := repo.CommitObject(ref.Hash())
 	if err != nil {
-		return "", plumbing.ZeroHash, err
+		return &RSLEntry{}, err
 	}
 
-	messageSplit := strings.Split(commitObj.Message, ":")
+	return parseRSLEntryMessage(commitObj.Message), nil
+}
 
-	return strings.Trim(messageSplit[0], " "),
-		plumbing.NewHash(strings.Trim(messageSplit[1], " ")), nil
+// GetEntry returns the entry corresponding to entryID.
+// TODO: There is no information yet about the signature for the entry.
+func GetEntry(entryID plumbing.Hash) (*RSLEntry, error) {
+	repo, err := common.GetRepositoryHandler()
+	if err != nil {
+		return &RSLEntry{}, err
+	}
+
+	ref, err := repo.Reference(plumbing.ReferenceName(RSLRef), true)
+	if err != nil {
+		return &RSLEntry{}, err
+	}
+
+	iteratorHash := ref.Hash()
+	for {
+		commitObj, err := repo.CommitObject(iteratorHash)
+		if err != nil {
+			return &RSLEntry{}, err
+		}
+
+		if iteratorHash == entryID {
+			return parseRSLEntryMessage(commitObj.Message), nil
+		}
+
+		if len(commitObj.ParentHashes) == 0 {
+			return &RSLEntry{}, ErrRSLEntryNotFound
+		}
+
+		if len(commitObj.ParentHashes) > 1 {
+			return &RSLEntry{}, ErrRSLBranchDetected
+		}
+
+		iteratorHash = commitObj.ParentHashes[0]
+	}
+}
+
+func parseRSLEntryMessage(message string) *RSLEntry {
+	entry := &RSLEntry{
+		Revoke: []plumbing.Hash{},
+	}
+
+	for _, l := range strings.Split(message, "\n") {
+		ls := strings.Split(l, ":")
+		switch strings.TrimSpace(ls[0]) {
+		case "ref":
+			entry.RefName = strings.TrimSpace(ls[1])
```

Comment:
You'll want to check the len - this will panic you don't get an actual split.

I also don't know if `:` is valid for ref names - you may need `ls[1:]` and `TrimPrefix`

Reply:
`fatal: 'new:branch' is not a valid branch name`

Luckily not as `:` is also used for `local:remote` mapping.

> You'll want to check the len - this will panic you don't get an actual split.

Good call, I'll add that.

Reply:
[3fbfb93](https://github.com/adityasaky/gittuf/pull/24/commits/3fbfb93fd65b7ccd8983df4e5c8ceffe41d7a021)